vyper/venom/effects.py

```diff
@@ -0,0 +1,59 @@
+from enum import Flag, auto
+
+
+class Effects(Flag):
+    STORAGE = auto()
+    TRANSIENT = auto()
+    MEMORY = auto()
+    IMMUTABLES = auto()
+    BALANCE = auto()
+    RETURNDATA = auto()
+
+
+EMPTY = Effects(0)
+ALL = ~EMPTY
+STORAGE = Effects.STORAGE
+TRANSIENT = Effects.TRANSIENT
+MEMORY = Effects.MEMORY
+IMMUTABLES = Effects.IMMUTABLES
+BALANCE = Effects.BALANCE
+RETURNDATA = Effects.RETURNDATA
+
+
+writes = {
+    "sstore": STORAGE,
+    "tstore": TRANSIENT,
+    "mstore": MEMORY,
+    "istore": IMMUTABLES,
+    "call": ALL,
+    "delegatecall": ALL,
+    "staticcall": MEMORY | RETURNDATA,
+    "create": ALL,
+    "create2": ALL,
+    "invoke": ALL,  # could be smarter, look up the effects of the invoked function
+    "dloadbytes": MEMORY,
+    "returndatacopy": MEMORY,
+    "calldatacopy": MEMORY,
+    "codecopy": MEMORY,
+    "extcodecopy": MEMORY,
+    "mcopy": MEMORY,
+}
+
+reads = {
+    "sload": STORAGE,
+    "tload": TRANSIENT,
+    "iload": IMMUTABLES,
+    "mload": MEMORY,
+    "mcopy": MEMORY,
+    "call": ALL,
+    "delegatecall": ALL,
+    "staticcall": ALL,
+    "returndatasize": RETURNDATA,
+    "returndatacopy": RETURNDATA,
+    "balance": BALANCE,
+    "selfbalance": BALANCE,
+    "log": MEMORY,
+    "revert": MEMORY,
+    "return": MEMORY,
```

Comment:
return has an offset to memory, right? so in theory you might be forced to zero-extend the memory

Reply:
applies to other opcodes as well

Reply:
do we really care about zero extension? i think gas is an effect we don't need to preserve

Reply:
what about `msize`?

Reply:
hmm that's true. but return can't be reordered past other instructions anyways

Reply:
you have the same for eg `mload`